### ISSUE TITLE
Wrong values used for composite primary key

### ISSUE THREAD

**Issue by OndrejPiza** [dotnet/efcore#18862]
Hi,

I'm not sure if this is bug or something is hidden somewhere in configuration and I missed it. 

I have little bit unusual structure. I have main table Player with join table FriendPlayer, which should consists of only two columns, PlayerId and FriendId, both columns are pointing to the Player table.

```C#
public class Player
{
      public Guid Id;

      public virtual ICollection<FriendPlayer> Friends { get; set; } = new List<FriendPlayer>();
}

public class FriendPlayer
    {
        /// <summary>
        /// Requesting side Id of friendship
        /// </summary>
        public Guid PlayerId { get; set; }
        
        /// <summary>
        /// Requesting side of friendship
        /// </summary>
        public Player Player { get; set; }

        /// <summary>
        /// Accepted side Id of friendship
        /// </summary>
        public Guid FriendId { get; set; }

        /// <summary>
        /// Accepted side of friendship
        /// </summary>
        public Player Friend { get; set; }
    }

class PlayerContext
{
  protected override void OnModelCreating(ModelBuilder modelBuilder)
  {
      ......
      modelBuilder.Entity<FriendPlayer>()
                .HasOne(fp => fp.Friend)
                .WithMany(p => p.Friends)
                .HasForeignKey(fp => fp.FriendId);
      modelBuilder.Entity<FriendPlayer>().HasKey(fp => new { fp.PlayerId, fp.FriendId });
  }

  public DbSet<Player> Players { get; set; } 

  public DbSet<FriendPlayer> FriendPlayers { get; set; }
}
```

First of all, I don't know if this approach is correct but I couldn't find any docs about this self reference same table from two properties to create one level of hierarchy.

Something must be wrong because when saving to DB, it uses PlayerId as both parts of composite primary key as well as values for columns, so basically I end up with same value in both columns. Did I miss something?

Thank you for any advice.

**Comment by ajcvickers** [comment]
@OndrejPiza Thanks for filing this issue. We will discuss in our next triage meeting.

**Comment by OndrejPiza** [comment]
@ajcvickers Thank you. I've did some try and debug and found out, that this is done via adding this relation object to parent collection. When I added it directly through context, it worked OK.

So when I attach this object to parent collection, during save this parent Id is propagated to all children entities based on Type of navigation property on each children, right?

Then it works "as supposed" I guess, only for this specific scenario it overwrites my other navigation property of the same type. 
Then I've found some article on SO, where it was describing how to achieve that, at least it works also by adding it to the collection, but I don't know if this is correct approach.

```
            builder.Entity<FriendPlayer>()
                .HasOne(fp => fp.Friend)
                .WithMany()
                .HasForeignKey(fp => fp.FriendId)
                .OnDelete(DeleteBehavior.Restrict);

            builder.Entity<FriendPlayer>()
                .HasOne(fp => fp.Player)
                .WithMany(p => p.FriendPlayers)
                .HasForeignKey(fp => fp.PlayerId);
```

https://stackoverflow.com/questions/49214748/many-to-many-self-referencing-relationship

Basically you have to have either two properties on parent for each, in my case I have another 3 relations like that so it would end up in 8 properties in total, or you have to leave one .WithMany() without specifying relationship.

**Comment by ajcvickers** [comment]
@OndrejPiza Answer on Stack Overflow looks correct.

**Comment by OndrejPiza** [comment]
It's is, but you end up with one unmapped navigational property, depending which one you choose not to have .WithMany() with corresponding collection. I can live with only ID being mapped, without loaded object, it's just something you have to consider, eg. doing null checking if you're converting whole object in some way or use your own loading logic. Since it is already part of the parent, I think it's not such a problem.

But I think that this is not quite a way how would I suppose or expect it to work, especially based on example in documentation. Could you please consider extending documentation about this special case?

**Comment by ajcvickers** [comment]
@OndrejPiza There can be zero, one, or two navigation properties for each relationship. However, the same navigation property cannot be used for two different relationships.
